Reimplement sequence elements

timonegk opened the pull request:
In this pull request, sequence elements are reimplemented similar to the AbstractAction- and AbstractDecisionElements. This eases the handling of sequence elements in the DSD and allows for better visualization. 
It closes #13 and closes #12.

timonegk committed:
Dynamic Stack Decider: reimplement sequence elements

Now, they are implemented like AbstractActionElements and
AbstractDecisionElements. This makes the implementation and debug data
generation easier.

timonegk committed:
Dynamic Stack Decider: make tests usable without xmlrunner

timonegk committed:
Dynamic Stack Decider Visualization: fix sequence elements

timonegk committed:
Visualization: show DSD parameters

Closes #11

timonegk commented:
In 180ce2964c3d0b86e1bece6ae2cfe11ec86b43b5, I also added the DSD parameters to the  (#11). Therefore, I switched the shapes of actions and decisions (actions are now boxes and decisions are ellipses). This looked better to me because the sequence elements now contain line breaks.

lilioid commented:
> In [180ce29](https://github.com/bit-bots/dynamic_stack_decider/commit/180ce2964c3d0b86e1bece6ae2cfe11ec86b43b5), I also added the DSD parameters to the (#11). Therefore, I switched the shapes of actions and decisions (actions are now boxes and decisions are ellipses). This looked better to me because the sequence elements now contain line breaks.

As I remember it @SammyRamone decided on the original shape of action and decision elements. He was quite sure on why they were shaped like they were

lilioid commented on dynamic_stack_decider/src/dynamic_stack_decider/sequence_element.py:
```diff
@@ -0,0 +1,58 @@
+        Pop a single element of the sequence
+        :return: True when the pop was successful, False when the end of the sequence has been reached
+        """
+        self.current_action_index += 1
```
Because this  statement is not encapsulated by your `if` block, consecutive calls to `pop_one()` would result in a weird `current_action_index` value.
Although it does not make sense to just call `pop_one()` a lot of times, I would still prefer the index incrementation to be protected

timonegk replied:
Fixed by using ´>=´ in the condition below

lilioid commented on dynamic_stack_decider/src/dynamic_stack_decider/sequence_element.py:
```diff
@@ -0,0 +1,58 @@
+
+        :rtype: dict
+        """
+        self.clear_debug_data()
```
Why is the debug data cleared before being sent?

I understand that you are setting it in the following lines but this is not an intuitive behavior. It would probably be better to clear it after sending.

timonegk replied:
Done in 5925e7345efb4e2e3a6020652d53cbb3e9e75105

lilioid commented on dynamic_stack_decider/tests/test_parser.py:
```diff
@@ -98,7 +97,11 @@ def test_multiple_subbehavior_references(self):
+                testRunner=xmlrunner.XMLTestRunner(output=output)
+            )
+    except ImportError:
+        unittest.main()
```
This is not strictly speaking part of the stack element reimplementation but I think we can let it pass ;)

timonegk replied:
Thank you :)

lilioid reviewed (changes requested):
In general I like the reimplementation and think it is a good addition to keeping the codebase clean.

None the less I have a few change requests for you to look over

ftsell commented:
Can you please wait a little before merging this PR because  I want to see why Jenkins is not adding test results?

lilioid commented:
Can you please wait a little before merging this PR?
I want to figure out why Jenkins is adding test results to the commits but not the PR.

timonegk committed:
dynamic stack decider: remove debug prints

timonegk committed:
dynamic stack decider: fix requested changes

lilioid committed:
ci: use correct docker container

previously jenkins was using the wrong docker container to run its test
in. This is now fixed

lilioid committed:
ci: add python-coverage as test dependency

lilioid committed:
ci: fix jenkins calling wrong coverage binary

the binary is not named `coverage` but rather `python-coverage`

lilioid committed:
ci: index dependencies from src directory

previously we were installing "all" dependencies. Because that did
not produce reliable results we are now installing all dependencies
needed by packages in the src directory

SammyRamone commented:
I'm okay with the code. As for the shapes @MrSunday334 maybe wants to check if he is fine with it, since we wrote it differently in the paper. I don't remember why we chose the other shapes thou (maybe due to similarity with behavior or decision trees?).

timonegk committed:
add default pylintrc but without no-else-return

SammyRamone commented:
after some thought, I decided that the shapes are fine, since there seems to be no real convention on this. please merge
